### DISCUSSION TITLE
Store asset transformations for reference when serving assets from outside of the studio app

### DISCUSSION THREAD

**Discussion by that1matt** [directus/directus#24064] [Feature Requests]
### Summary

When not using the Directus Studio app to serve assets, there is currently no way to track asset transformations through the API or through Directus system collections.

We deliver our assets through our own CDN. We currently use Directus Studio app as our asset management system, where Directus stores all the assets on our own S3 bucket. We then use Akamai to serve those assets on our sites through a custom URL. This limits our ability to use asset transformations in a programatic way when serving our assets.

Currently the only way to see these transformations is S3.

Here is any example
```shell
2024-04-03 10:35:36      16489 f8cf5f77-bd52-451f-968d-8ecf889bd9ba.png
2024-04-03 10:35:36      33674 f8cf5f77-bd52-451f-968d-8ecf889bd9ba__4f7f2791732f0b8f33c44abe69d6e5b615ea3f17.png
2024-11-05 13:08:40      11182 f8cf5f77-bd52-451f-968d-8ecf889bd9ba__7abd30a2f63f675625d6888e2ed9d65992751e3c.avif
2024-11-05 13:09:00      18895 f8cf5f77-bd52-451f-968d-8ecf889bd9ba__da7fd914f287cf8bb6fa70d605fa544b0ae8e081.png
2024-11-05 13:13:56       1593 f8cf5f77-bd52-451f-968d-8ecf889bd9ba__e9dcbd6d37cf1a1351d2786febd96ceaf3ef6132.avif
2024-04-03 12:46:49       6485 f8cf5f77-bd52-451f-968d-8ecf889bd9ba__f15aa4b815722867859d2a3fa46ab50834dfc8dc.avif
```

### Motivation

Be able to use Directus Studio app features when using an external CDN provider to serve assets managed in Directus.

### Alter `directus_files`
- Altering `directus_files` to include new fields to track asset transformations and the properties used to transform them
- Potentially a field named `asset_transformation_filename` that would hold the filename of the asset transformation.
- Potentially a field named `asset_transformation_options` that returns a JSON of the arguments used to transform the asset. This would allow the ability to filter multiple asset transformation for the same file, and return the one with a specific set of transformation options.

### Create a new `directus_files_transformations` collection
- Create a new collection to track the asset transformation
- Include fields for each transformation option, or a JSON object of the transformation options
- Include the asset transformation name

### API
- The API object for files should return a reference to the transformed asset, along with the transformation options.

### Must Have:
- Ability to reference the transformed assets through the API
- Ability to reference the transformation options through the API
- Ability to show the original asset, along with all the transformed assets

### Could Have:
- Versioning of the transformed assets

### Drawbacks

- Refactoring the `directus_files` can cause potential issues of current implementations.
- Would require updates to all the asset connectors for each storage driver
- Adds additional queries if using a junction table
- Adds additional fields to files object

### Alternatives

- Could build a filter to query the S3 bucket by the file `id` prefix, and return all transformations, then some additional logic to return the correct transformation.
- Only allow 1 transformation per file, and then filter through the S3 bucket.

### Adoption Strategy

Ideally, this wouldn't be a breaking change, but would be a minor change that adds new features, without affecting current functionality.

### Unresolved Questions

- What is the best option?
- What option limits the potential for issues?

**Comment by rijkvanzanten** [comment]
Just to make sure I understand the usage flow here; you'd want to get a list of already-generated-thumbnails from the /files endpoint so you know whether you can hit your CDN directly, or hit Directus if the expected on doesn't exist yet?

If reading files directly from the CDN is the goal, would a "Generate all presets on upload" + a way to know how to programmatically generate the same filepaths achieve the same goal? 🤔

**Reply by that1matt** [reply]
> Just to make sure I understand the usage flow here; you'd want to get a list of already-generated-thumbnails from the /files endpoint so you know whether you can hit your CDN directly, or hit Directus if the expected on doesn't exist yet?

Yes, hitting the `/files` endpoint and getting a list of current transformations along with their params for each transformation, otherwise use the API to create a new asset transformation. I believe the [requesting-a-thumbnail](https://docs.directus.io/reference/files.html#requesting-a-thumbnail) has this functionality. 

> If reading files directly from the CDN is the goal, would a "Generate all presets on upload" + a way to know how to programmatically generate the same filepaths achieve the same goal? 

I would not want to limit it to `Generating all presets on upload` as that would cause additional API calls to get the asset, then re-upload to get a new asset transformation. 

Having a way to programmatically generate the same file transformations, would also be beneficial.

I think ultimately right now within Directus, there is no way to determine what files have transformations, and what properties are set when that transformation was created.

Directus would be the source of truth in managing all assets and their transformations, but have a way to deliver those assets from our own CDN.

**Reply by rijkvanzanten** [reply]
Just (divergently) thinking out loud here, would a 'best case' scenario here be that you'd hit the /assets endpoint of Directus using the transforms you want, and it 301 redirects to the CDN provider if it exists, or generates it on the fly and then 301s to the CDN link for the asset dynamically? 🤔

**Reply by that1matt** [reply]
For our use case, hitting the `/assets` endpoint to create the transformations and storing it in the S3 bucket as it does now. Just providing a reference to the transformation (and transformation options) so we can handle the logic inside our app. Since our CDN serves the assets directly from the S3 bucket using the file `id`.

**Comment by ampsu** [comment]
Did you found an okay solution for that? I currently bumped into the same issue. Im using S3/R2 bucket and images are transformed by Directus to multiple sizes but there is no good way to get filenames for those transformed images. `d2283d02-2e70-4b94-9673-f51dd99f8723__7abd30a2f63f675625d6888e2ed9d65992751e3c.avif` isnt helpful when we cant retrieve that name through collection API responses.
